lib/Dialect/Substrait/Transforms/EmitDeduplication.cpp

```diff
@@ -116,6 +116,60 @@ createDeduplicatingEmit(Value input, SmallVector<int64_t> &reverseMapping,
   return {newEmitOp, newInputMapping.size(), true};
 }
 
+/// Deduplicates the fields of the region with a single `Tuple` argument using
+/// the provided (deduplicating) mapping. This involves changing the type of the
+/// region argument to the provided `newElementType`, which must be the type
+/// obtained by applying deduplication to the argument type of the provided
+/// `region`, as well as changing all `field_reference` ops using the region
+/// argument to work on the deduplicated type.
+// TODO(ingomueller): We could add an overload for this function that computes
+// `newElementType` from the type of the region argument and the mapping.
+void deduplicateRegionArgs(Region &region, ArrayAttr newMapping,
+                           Type newElementType, PatternRewriter &rewriter) {
+  assert(region.getNumArguments() == 1 &&
+         "only regions with 1 argument are supported");
+  auto oldElementType = cast<TupleType>(region.getArgument(0).getType());
+  int64_t numOldFields = oldElementType.getTypes().size();
+
+  // For each position in the original input type, compute which position it
+  // corresponds to in the deduplicated input. This is required for replacing
+  // field references to the original type with references to the deduplicated
+  // type.
+  SmallVector<int64_t> oldToNewPositions;
+  oldToNewPositions.reserve(numOldFields);
+  {
+    llvm::DenseMap<int64_t, int64_t> indexPositions;
+    for (auto attr : newMapping.getAsRange<IntegerAttr>()) {
+      int64_t index = attr.getInt();
+      int64_t pos = indexPositions.size();
+      auto [it, success] = indexPositions.try_emplace(index, pos);
+      oldToNewPositions.push_back(it->second);
+    }
+  }
+
+  // Update field references using the region argument.
+  for (Operation *user : region.getArgument(0).getUsers()) {
+    // We are only interested in `field_reference` ops.
+    if (!isa<FieldReferenceOp>(user))
```

Comment:
dyn_cast instead of isa followed by cast

Reply:
Done!